pkg/sql/reassign_owned_by.go

```diff
@@ -13,15 +13,114 @@ package sql
 import (
 	"context"
 
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/dbdesc"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/util/errorutil/unimplemented"
 )
 
 // ReassignOwnedByNode represents a REASSIGN OWNED BY <name> TO <name> statement.
 // TODO(angelaw): to implement
-//type reassignOwnedByNode struct {
-//}
+type reassignOwnedByNode struct {
+	n *tree.ReassignOwnedBy
+}
 
 func (p *planner) ReassignOwnedBy(ctx context.Context, n *tree.ReassignOwnedBy) (planNode, error) {
+
+	// Check all roles in old roles exist
+	// And that current user has membership on old role (PostgresQL requirement)
+	// TODO(angelaw): Refactor the error checking out to authorization?
+	currentUser := p.SessionData().User
+
+	memberOf, err := p.MemberOfWithAdminOption(ctx, currentUser)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, oldRole := range n.OldRoles {
+		roleExists, err := p.RoleExists(ctx, oldRole)
+		if err != nil {
+			return nil, err
+		}
+
+		if !roleExists {
+			return nil, pgerror.Newf(pgcode.UndefinedObject, "role/user %q does not exist", oldRole)
+		}
+
+		if _, ok := memberOf[oldRole]; !ok {
+			return nil, pgerror.Newf(
+				// TODO(angelaw): find error message from psql
+				pgcode.InsufficientPrivilege, "current user must be member of role %q", oldRole)
+		}
+	}
+
+	// Check new role exists
+	roleExists, err := p.RoleExists(ctx, n.NewRole)
+	if err != nil {
+		return nil, err
+	}
+
+	if !roleExists {
+		return nil, pgerror.Newf(pgcode.UndefinedObject, "role/user %q does not exist", n.NewRole)
+	}
+
+	// And that current user has membership on new role (PostgresQL requirement)
+	if _, ok := memberOf[n.NewRole]; !ok {
+		return nil, pgerror.Newf(
+			// TODO(angelaw): find error message from psql
+			pgcode.InsufficientPrivilege, "current user must be member of role %q", n.NewRole)
+	}
+
+	// Find all objects in current database - use getAllDescriptors, newInternalLookupCtxFromDescriptors
+	allDescs, err := p.Descriptors().GetAllDescriptors(ctx, p.txn)
+	if err != nil {
+		return nil, err
+	}
+
+	currentDatabase := p.CurrentDatabase()
+	currentDbDesc, err := p.ResolveMutableDatabaseDescriptor(ctx, currentDatabase, true)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(angelaw): This is messy but we need an immutableDbDesc for description and
+	// we only have a ResolveMutableDatabaseDescriptor — any better ideas?
+	lCtx := newInternalLookupCtx(ctx, allDescs, currentDbDesc.ImmutableCopy().(*dbdesc.Immutable))
+
+	// For each role in old roles
+	for _, oldRole := range n.OldRoles {
+		// Iterate through each object in current DB
+		for _, dbDesc := range lCtx.dbDescs {
+			// We only expect 1 matching db (current database)
+			if p.IsOwner(dbDesc, oldRole) {
+				p.AlterDatabaseOwner(ctx, tree.AlterDatabaseOwner) // Does not include new owner string
```

Comment:
Calling all of these methods, `AlterDatabaseOwner` `alterSchemaOwner, etc are not as straightforward as they are in psql. Each takes in different arguments that I will need to populate individually. 

Does this seem like the right path to keep moving towards, or should I look to directly change the ownership here without calling these methods?